File: apps/starter-ios/StarterApp/Core/Analytics/AnalyticsService.swift
import Foundation

struct AnalyticsEvent: Encodable {
    let name: String
    let category: String?
    let properties: [String: String]?
}

struct AnalyticsSummary: Codable, Equatable {
    struct TopEvent: Codable, Equatable {
        let name: String
        let count: Int
    }

    let totalEvents: Int
    let uniqueEvents: Int
    let topEvents: [TopEvent]
    let lastEventAt: String?
}

struct AnalyticsService {
    func track(name: String, category: String? = nil, properties: [String: String]? = nil) async throws {
        let _: TrackAcceptedResponse = try await APIClient.shared.request(
            "analytics/events",
            method: "POST",
            body: AnalyticsEvent(name: name, category: category, properties: properties)
        )
    }

    func trackAuthenticated(
        name: String,
        category: String? = nil,
        properties: [String: String]? = nil
    ) async throws {
        let _: TrackAcceptedResponse = try await APIClient.shared.request(
            "analytics/events/authenticated",
            method: "POST",
            body: AnalyticsEvent(name: name, category: category, properties: properties)
        )
    }

    func fetchSummary() async throws -> AnalyticsSummary {
        try await APIClient.shared.request("analytics/summary")
    }
}

private struct TrackAcceptedResponse: Decodable {
    let accepted: Bool
}

File: apps/starter-android/app/src/main/kotlin/com/onlystack/starterapp/core/analytics/AnalyticsService.kt
package com.onlystack.starterapp.core.analytics

import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AnalyticsService @Inject constructor(
    private val apiService: AnalyticsApiService,
) {
    suspend fun track(
        name: String,
        category: String? = null,
        properties: Map<String, String>? = null,
    ): Result<Unit> = runCatching {
        apiService.track(AnalyticsEvent(name = name, category = category, properties = properties))
        Unit
    }

    suspend fun trackAuthenticated(
        name: String,
        category: String? = null,
        properties: Map<String, String>? = null,
    ): Result<Unit> = runCatching {
        apiService.trackAuthenticated(
            AnalyticsEvent(name = name, category = category, properties = properties)
        )
        Unit
    }

    suspend fun fetchSummary(): Result<AnalyticsSummary> = runCatching {
        apiService.summary()
    }
}
